src/halmos/bytevec.py

```diff
@@ -0,0 +1,754 @@
+        return Chunk._empty
+
+    def __iter__(self):
+        raise TypeError("Chunk object is not iterable")
+
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            start = key.start or 0
+            stop = key.stop if key.stop is not None else self.length
+            step = key.step or 1
+
+            if step != 1:
+                raise NotImplementedError(f"slice with step={step} not supported")
+
+            # TODO: should we return an empty chunk or raise if start >= stop?
+            # if start >= stop or start >= self.length:
+            #     return Chunk.empty()
+
+            if not (0 <= start <= self.length) or not (0 <= stop <= self.length):
+                raise IndexError(key)
+
+            return self.slice(start, stop)
+
+        return self.get_byte(key)
+
+    def __len__(self):
+        return self.length
+
+    @abstractmethod
+    def get_byte(self, offset):
+        raise NotImplementedError
+
+    @abstractmethod
+    def slice(self, start, stop) -> "Chunk":
+        raise NotImplementedError
+
+    @abstractmethod
+    def unwrap(self) -> UnionType[bytes, BitVecRef]:
+        raise NotImplementedError
+
+
+class ConcreteChunk(Chunk):
+    """A chunk of concrete, native bytes"""
+
+    def __init__(self, data: bytes, start=0, length=None):
+        self.data_byte_length = len(data)
+
+        if length is None:
+            length = self.data_byte_length - start
+
+        assert start >= 0
+        assert length >= 0
+        assert start + length <= self.data_byte_length
+
+        super().__init__(data, start, length)
+
+    def get_byte(self, offset) -> int:
+        """
+        Return a single byte at the given offset (fast given that this is a native operation)
+
+        Complexity: O(1)
+        """
+
+        if not (0 <= offset < self.length):
+            raise IndexError(offset)
+
+        return self.data[self.start + offset]
+
+    def slice(self, start, stop) -> "ConcreteChunk":
+        """
+        Return a slice of the chunk, as a new chunk (this is fast and does not copy the backing data)
+
+        Complexity: O(1)
+        """
+
+        return ConcreteChunk(self.data, self.start + start, stop - start)
+
+    def unwrap(self) -> bytes:
+        """
+        Return the data as a single value, this is where the actual slicing/copying is done
+
+        Complexity: O(n)
+        """
+
+        # if this chunk represents the entire data, return the data itself
+        if self.length == self.data_byte_length:
+            return self.data
+
+        return self.data[self.start : self.start + self.length]
+
+    def __eq__(self, other):
+        # allow comparison of empty symbolic and concrete chunks
+        if isinstance(other, Chunk) and not self and not other:
+            return True
+
+        if not isinstance(other, ConcreteChunk):
+            return False
+
+        return self.length == len(other) and self.unwrap() == other.unwrap()
```

Comment:
good point, I'll add a comment

note that since we expect the backing type to be native bytes of modest sizes (< 100kB), I suspect letting the interpreter do the slicing+comparison will be faster than any looping we would do ourselves